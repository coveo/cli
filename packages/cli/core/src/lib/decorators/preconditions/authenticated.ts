import {
  PreconditionError,
  PreconditionErrorCategory,
} from '../../errors/preconditionError';
import {
  AuthenticationStatus,
  getAuthenticationStatus,
} from '../../platform/authenticatedClient';

export function IsAuthenticated() {
  return async function () {
    const status = await getAuthenticationStatus();
    if (status === AuthenticationStatus.LOGGED_OUT) {
      throw new PreconditionError(
        'Not currently logged in. Run coveo auth:login first.',
        {category: PreconditionErrorCategory.Authentication}
      );
    }

    if (status === AuthenticationStatus.EXPIRED) {
      throw new PreconditionError(
        'Authentication token is expired. Run coveo auth:login first.',
        {category: PreconditionErrorCategory.Authentication}
      );
    }
  };
}
