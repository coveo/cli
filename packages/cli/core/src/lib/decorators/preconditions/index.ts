// TODO: Remove, unusued
export * from './git';
// TODO: Move to `@coveo/plugin-cli-ui
export * from './node';
// TODO: Move to `@coveo/plugin-cli-ui
export * from './npx';
// TODO: Move to `@coveo/plugin-cli-ui
export * from './npm';
// TODO: Move to `@coveo/plugin-cli-ui
export * from './ng';
