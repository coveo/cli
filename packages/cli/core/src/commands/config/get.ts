import {CLICommand} from '@coveo/cli-commons/command/cliCommand';
import {Config} from '@coveo/cli-commons/config/config';
import {ConfigRenderer} from '@coveo/cli-commons/config/configRenderer';
import {Trackable} from '@coveo/cli-commons/preconditions/trackable';
import type {Example} from '@oclif/core/lib/interfaces';

export default class Get extends CLICommand {
  public static description = 'Display the current Coveo CLI configuration.';

  public static args = [
    {
      name: 'key',
      description: 'The config key for which to show the value',
      required: false,
    },
  ];

  public static examples: Example[] = [
    {
      command: 'coveo config:get',
      description: 'Get all the configuration values',
    },
    {
      command: 'coveo config:get organization',
      description: 'Get the organization to which you are connected',
    },
    {
      command: 'coveo config:get accessToken',
      description: 'Get the access token given to you by the Coveo Platform',
    },
  ];

  @Trackable()
  public async run() {
    const {args} = await this.parse(Get);
    const cfg = new Config(this.config.configDir);
    const keysToRender = args.key ? [args.key] : undefined;
    ConfigRenderer.render(cfg, keysToRender);
  }
}
