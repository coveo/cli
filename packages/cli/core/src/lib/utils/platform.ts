import {Config} from '@coveo/cli-commons/config/config';

export function getTargetOrg(config: Config, target?: string) {
  if (target) {
    return target;
  }
  const cfg = config.get();
  return cfg.organization;
}
