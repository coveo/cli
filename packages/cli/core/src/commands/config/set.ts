import {CLICommand} from '@coveo/cli-commons/command/cliCommand';
import {CliUx, Flags} from '@oclif/core';
import {Config} from '@coveo/cli-commons/config/config';
import {AuthenticatedClient} from '@coveo/cli-commons/platform/authenticatedClient';
import {
  IsAuthenticated,
  Preconditions,
} from '@coveo/cli-commons/preconditions/index';
import {Trackable} from '@coveo/cli-commons/preconditions/trackable';
import {InvalidCommandError} from '../../lib/errors/InvalidCommandError';
import {ConfigRenderer} from '@coveo/cli-commons/config/configRenderer';
import type {Example} from '@oclif/core/lib/interfaces';

export default class Set extends CLICommand {
  public static description = 'Modify the current Coveo CLI configuration.';

  public static flags = {
    // TODO CDX-1246
    environment: Flags.string({
      char: 'e',
      hidden: true,
    }),
    // TODO CDX-1246
    region: Flags.string({
      char: 'r',
      hidden: true,
    }),
    organization: Flags.string({
      char: 'o',
      description:
        'The identifier of the organization inside which to perform operations. See <https://docs.coveo.com/en/1562/#organization-id-and-other-information>.',
      helpValue: 'myOrgID',
    }),
  };

  public static examples: Example[] = [
    {
      command: 'coveo config:set --organization myOrgId',
      description: 'connect to a different organization',
    },
  ];

  @Trackable()
  @Preconditions(IsAuthenticated())
  public async run() {
    const {flags} = await this.parse(Set);
    // TODO CDX-1246
    if (flags.environment || flags.region) {
      CliUx.ux.error(
        'To connect to a different region or environment, use the `auth:login` command'
      );
    }
    if (Object.entries(flags).length === 0) {
      throw new InvalidCommandError('Command should contain at least 1 flag');
    }
    const cfg = new Config(this.config.configDir);
    if (flags.organization) {
      await this.verifyOrganization(flags.organization);
      cfg.set('organization', flags.organization);
    }

    ConfigRenderer.render(cfg);
  }

  private async verifyOrganization(org: string) {
    const hasAccess = await new AuthenticatedClient().getUserHasAccessToOrg(
      org
    );
    if (!hasAccess) {
      this.error(
        `You either don't have access to organization ${org}, or it doesn't exist.`
      );
    }
  }
}
