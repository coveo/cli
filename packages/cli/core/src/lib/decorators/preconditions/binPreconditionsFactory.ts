import type {Command} from '@oclif/core';
import {dedent} from 'ts-dedent';
import {spawnProcessOutput, SpawnProcessOutput} from '../../utils/process';
import {satisfies, validRange} from 'semver';
import {
  PreconditionError,
  PreconditionErrorCategory,
} from '../../errors/preconditionError';

export interface BinPreconditionsOptions {
  params?: Array<string>;
  prettyName: string;
  installLink?: string;
  howToInstallBinText?: string;
}

const defaultOptions: Required<Omit<BinPreconditionsOptions, 'prettyName'>> = {
  params: ['--version'],
  installLink:
    'https://github.com/coveo/cli/wiki/Node.js,-NPM-and-NPX-requirements',
  howToInstallBinText: '',
};

export function getBinVersionPrecondition(
  binaryName: string,
  options: BinPreconditionsOptions,
  versionExtractor: (stdout: string) => string = (stdout) => stdout
) {
  const appliedOptions: Required<BinPreconditionsOptions> = {
    ...defaultOptions,
    ...options,
    prettyName: options.prettyName,
  };
  return function (versionRange: string) {
    return async function (target: Command) {
      if (!validRange(versionRange)) {
        const message = dedent`
          Required version invalid: "${versionRange}".
          Please report this error to Coveo: https://github.com/coveo/cli/issues/new
        `;
        throw new PreconditionError(message, {
          category: PreconditionErrorCategory.InvalidBinVersionRange,
        });
      }

      const output = await spawnProcessOutput(
        binaryName,
        appliedOptions.params
      );

      await checkIfBinIsInstalled(target, binaryName, appliedOptions, output);
      const version = versionExtractor(output.stdout);

      if (!satisfies(version, versionRange)) {
        const message = dedent`
          ${target.id} needs a ${
          appliedOptions.prettyName
        } version in this range: "${versionRange}"
          Version detected: ${version}

          ${warnHowToInstallBin(appliedOptions)}
        `;
        throw new PreconditionError(message, {
          category: PreconditionErrorCategory.InvalidBinVersionRange,
        });
      }
    };
  };
}

export function getBinInstalledPrecondition(
  binaryName: string,
  options: BinPreconditionsOptions
) {
  const appliedOptions: Required<BinPreconditionsOptions> = {
    ...defaultOptions,
    ...options,
    prettyName: options.prettyName,
  };
  return function () {
    return async function (target: Command) {
      const output = await spawnProcessOutput(
        binaryName,
        appliedOptions.params
      );
      await checkIfBinIsInstalled(target, binaryName, appliedOptions, output);
    };
  };
}

async function checkIfBinIsInstalled(
  target: Command,
  binaryName: string,
  options: Required<BinPreconditionsOptions>,
  output: SpawnProcessOutput
): Promise<void | never> {
  if (output.exitCode === 'ENOENT') {
    const warningMessage = dedent`${target.id} requires ${
      options.prettyName
    } to run.

    ${warnHowToInstallBin(options)}`;
    throw new PreconditionError(warningMessage, {
      category: PreconditionErrorCategory.MissingBin,
    });
  }

  if (output.exitCode !== '0') {
    const message = dedent`
      ${target.id} requires a valid ${options.prettyName} installation to run.
      An unknown error happened while running ${binaryName} ${options.params.join(
      ' '
    )}.
      ${output.stderr}

      ${warnHowToInstallBin(options)}
    `;
    throw new PreconditionError(message, {
      category: PreconditionErrorCategory.InvalidBinInstallation,
    });
  }
}

function warnHowToInstallBin(options: BinPreconditionsOptions) {
  const newParagraph = '\n\n';
  const howToInstallMessage = [];
  if (options.howToInstallBinText) {
    howToInstallMessage.push(options.howToInstallBinText);
  }
  howToInstallMessage.push(
    `Please visit ${options.installLink} for more detailed installation information.`
  );
  return howToInstallMessage.join(newParagraph);
}
