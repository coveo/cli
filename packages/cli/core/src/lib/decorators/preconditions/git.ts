import {getBinInstalledPrecondition} from './binPreconditionsFactory';

export const IsGitInstalled = getBinInstalledPrecondition('git', {
  prettyName: 'Git',
  installLink: 'https://git-scm.com/book/en/v2/Getting-Started-Installing-Git',
});
