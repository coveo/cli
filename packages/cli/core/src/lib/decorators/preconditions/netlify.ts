import {appendCmdIfWindows} from '../../utils/os';
import {getBinVersionPrecondition} from './binPreconditionsFactory';

const angularCliVersionMatcher = /^netlify-cli\/(?<version>\d+\.\d+\.\d+).*$/m;

export const IsNetlifyCliVersionInRange = getBinVersionPrecondition(
  appendCmdIfWindows`netlify`,
  {
    prettyName: 'Netlify-CLI',
    howToInstallBinText:
      'You can install the Netlify-CLI by running npm i -g netlify-cli',
    installLink: 'https://docs.netlify.com/cli/get-started/#installation',
    params: ['version'],
  },
  (stdout: string) =>
    stdout.match(angularCliVersionMatcher)?.groups?.['version'] ?? ''
);
