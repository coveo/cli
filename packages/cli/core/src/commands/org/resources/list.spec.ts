jest.mock('@coveo/cli-commons/config/config');
jest.mock('@coveo/cli-commons/preconditions/trackable');
jest.mock('@coveo/cli-commons/preconditions/authenticated');

jest.mock('@coveo/cli-commons/platform/authenticatedClient');

import {AuthenticatedClient} from '@coveo/cli-commons/platform/authenticatedClient';
import {test} from '@oclif/test';
import {ResourceSnapshotsModel} from '@coveo/platform-client';
import {Config} from '@coveo/cli-commons/config/config';
import {IsAuthenticated} from '@coveo/cli-commons/preconditions';
import {mockPreconditions} from '@coveo/cli-commons/preconditions/mockPreconditions';

const mockedAuthenticatedClient = jest.mocked(AuthenticatedClient);
const mockedConfig = jest.mocked(Config);
const mockedIsAuthenticated = jest.mocked(IsAuthenticated);

const doMockPreconditions = function () {
  const preconditionStatus = {
    authentication: true,
  };
  const mockedPreconditions = mockPreconditions(preconditionStatus);
  mockedIsAuthenticated.mockReturnValue(mockedPreconditions.authentication);
};

const mockResourceSnapshotModel: () => ResourceSnapshotsModel = () => ({
  createdBy: 'bob',
  createdDate: new Date(1487076708000).getTime(),
  id: 'id',
  targetId: 'targetId',
  contentSummary: {foo: 1, bar: 2},
  developerNote: 'some notes',
  originId: 'originId',
  reports: [],
  synchronizationReports: [],
});

const mockListSnapshots = jest
  .fn()
  .mockReturnValue(Promise.resolve([mockResourceSnapshotModel()]));

describe('org:resources:list', () => {
  beforeAll(() => {
    doMockPreconditions();
    mockedAuthenticatedClient.mockImplementation(
      () =>
        ({
          getClient: () =>
            Promise.resolve({resourceSnapshot: {list: mockListSnapshots}}),
        } as unknown as AuthenticatedClient)
    );
    mockedConfig.mockImplementation(
      () =>
        ({
          get: {
            organization: 'foo',
          },
        } as unknown as Config)
    );
  });

  test
    .do(() => {
      mockListSnapshots.mockReturnValueOnce(Promise.resolve([]));
    })
    .stdout()
    .stderr()
    .command(['org:resources:list', '-o', 'foo'])
    .it('works when there is no snapshot available', (ctx) => {
      expect(ctx.stdout).toContain(
        'There is no configuration snapshot available in organization foo'
      );
    });

  test
    .stdout()
    .stderr()
    .command(['org:resources:list', '-o', 'foo'])
    .it('print the available snapshot in a table', (ctx) => {
      expect(ctx.stdout).toMatchSnapshot();
    });

  test
    .stdout()
    .stderr()
    .command(['org:resources:list', '-o', 'foo', '--no-header'])
    .it('print the available snapshot in a table without header', (ctx) => {
      expect(ctx.stdout).toMatchSnapshot();
    });

  test
    .stdout()
    .stderr()
    .command(['org:resources:list', '-o', 'foo', '--output=json'])
    .it('print the available snapshot in a JSON format', (ctx) => {
      expect(ctx.stdout).toMatchSnapshot();
    });
});
