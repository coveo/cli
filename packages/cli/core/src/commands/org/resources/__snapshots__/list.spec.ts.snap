// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`org:resources:list print the available snapshot in a JSON format 1`] = `
"[
  {
    "id": "id",
    "createdBy": "bob",
    "createdDate": "2017-02-14T12:51:48.000Z",
    "targetId": "targetId",
    "developerNote": "some notes"
  }
]
"
`;

exports[`org:resources:list print the available snapshot in a table 1`] = `
" Id Created by Created date             Target id Developer note 
 ── ────────── ──────────────────────── ───────── ────────────── 
 id bob        2017-02-14T12:51:48.000Z targetId  some notes     
"
`;

exports[`org:resources:list print the available snapshot in a table without header 1`] = `
" id bob        2017-02-14T12:51:48.000Z targetId  some notes     
"
`;
